compiler/rustc_arena/src/lib.rs

```diff
@@ -114,6 +118,72 @@ impl<T> Default for TypedArena<T> {
     }
 }
 
+trait IterExt<T> {
+    fn alloc_from_iter(self, arena: &TypedArena<T>) -> &mut [T];
+}
+
+impl<I, T> IterExt<T> for I
+where
+    I: IntoIterator<Item = T>,
+{
+    #[inline]
+    default fn alloc_from_iter(self, arena: &TypedArena<T>) -> &mut [T] {
+        let vec: SmallVec<[_; 8]> = self.into_iter().collect();
+        vec.alloc_from_iter(arena)
+    }
+}
+
+impl<T, const N: usize> IterExt<T> for std::array::IntoIter<T, N> {
+    #[inline]
+    fn alloc_from_iter(self, arena: &TypedArena<T>) -> &mut [T] {
+        if N == 0 {
```

Comment:
I think this should be `self.len() == 0` or similar.

Reply:
The code below should work with a zero length, right? Unless alloc_from_slice has a nonzero precondition on the length which isn't documented...

Reply:
It should. My thought was that the iterator is not guaranteed to be on the first element, so `N == self.len()` isn't guaranteed. I shall doule check later, thoguh.

Reply:
Aha, no, the code below will not work -- alloc_raw_slice asserts a nonzero length. So we should compare against the length here.

Reply:
Fixed, and I've aligned the other places as well.